src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDag.cs

```diff
@@ -0,0 +1,248 @@
+
+        public HashSet<LabelSymbol> ReachableLabels
+        {
+            get
+            {
+                if (_reachableLabels == null)
+                {
+                    // Compute the set of reachable labels at the BoundDecision leaves.  We do this
+                    // iteratively rather than recursively so that we do not overflow the stack on a
+                    // large switch statement.
+                    var workQueue = ArrayBuilder<BoundDecisionDag>.GetInstance();
+                    var processed = PooledHashSet<BoundDecisionDag>.GetInstance();
+                    var reachableLabels = new HashSet<LabelSymbol>();
+                    workQueue.Add(this);
+                    processWorkQueue();
+                    workQueue.Free();
+                    processed.Free();
+                    _reachableLabels = reachableLabels;
+
+                    void processWorkQueue()
+                    {
+                        while (workQueue.Count != 0)
+                        {
+                            BoundDecisionDag node = workQueue.Pop();
+                            if (node == null || !processed.Add(node))
+                            {
+                                continue;
+                            }
+
+                            switch (node)
+                            {
+                                case BoundEvaluationPoint x:
+                                    workQueue.Push(x.Next);
+                                    break;
+                                case BoundDecisionPoint x:
+                                    workQueue.Push(x.WhenTrue);
+                                    workQueue.Push(x.WhenFalse);
+                                    break;
+                                case BoundWhenClause x:
+                                    workQueue.Push(x.WhenTrue);
+                                    workQueue.Push(x.WhenFalse); // possibly null, handled later
+                                    break;
+                                case BoundDecision x:
+                                    reachableLabels.Add(x.Label);
+                                    break;
+                                default:
+                                    throw ExceptionUtilities.UnexpectedValue(node.Kind);
+                            }
+                        }
+                    }
+                }
+
+                return _reachableLabels;
+            }
+        }
+
+        /// <summary>
+        /// Return a list of all the nodes reachable from this one, in a topologically sorted order.
+        /// </summary>
+        public ImmutableArray<BoundDecisionDag> TopologicallySortedNodes()
+        {
+            // We use an iterative topological sort to avoid overflowing the compiler's runtime stack for a large switch statement.
+            return TopologicalSort.IterativeSort<BoundDecisionDag>(new[] { this }, successorFunction);
```

Comment:
>successorFunction [](start = 83, length = 17)

Do we need `successorFunction` and `BoundDecisionDag.Successors()` above, or could this delegate be `d => d.Successors()`?